components-api/src/main/scala/pl/touk/nussknacker/engine/api/Service.scala

```diff
@@ -33,13 +33,52 @@ abstract class Service extends Lifecycle with Component
 // TODO: EagerService shouldn't extend Lifecycle, instead ServiceInvoker should extend it - see notice in ProcessCompilerData.lifecycle
 abstract class EagerService extends Service
 
-trait ServiceInvoker {
+trait ServiceLogic {
 
-  def invokeService(params: Map[String, Any])(
-      implicit ec: ExecutionContext,
+  def run(
+      paramsEvaluator: ParamsEvaluator
+  )(implicit runContext: RunContext, executionContext: ExecutionContext): Future[Any]
+
+}
+
+object ServiceLogic {
+
+  final case class RunContext(
       collector: InvocationCollectors.ServiceInvocationCollector,
       contextId: ContextId,
       componentUseCase: ComponentUseCase
-  ): Future[Any]
+  )
+
+  trait ParamsEvaluator {
+    def evaluate(additionalVariables: Map[String, Any] = Map.empty): EvaluatedParams
```

Comment:
My first implementation (available in the draft you saw) has sth like that. A developer could modify the context. But I started to wonder if it's ok to give him a possibility to modify eg. ContextId. So, I decided to limit it.  

The `Context` currently has 3 params:
1. `id`
2. `variables` 
3. `parentContext`

Because it's a case class, the `copy` method is available and we should consider it's ok to suggest that the `id` and the `parentContext` can be modified too. But maybe there is some value, I'm not aware about.